### PR TITLE
torch.jit.verify

### PR DESCRIPTION
**Pull request by ezyang**
A few notes about the implementation:
- Need to plumb 'devices' through to the 'fork_rng' calls.  You definitely
  want these; it makes verify run A LOT faster
- New keyword argument for compiled model execution, '_force_trace', which
  forces us to retrace a model.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

### CODE REVIEW

File: torch/jit/__init__.py [hunk 1]
```diff
@@ -580,5 +567,85 @@ def _time(trace_name, name, time=True):
...
+    # for this case
+    if not isinstance(model, _CompiledMixin):
+        raise TypeError("Cannot verify an uncompiled module.  Add @torch.jit.compile to compile it")
+
+    if not isinstance(args, tuple):
+        args = (args,)
+
+    verify_args = _clone_inputs(args)
+
+    verify_state = model.state_dict()
```
Comment (fmassa) [inline torch/jit/__init__.py]: I believe this doesn't perform a copy of the parameters / buffers, but just return references to it.
Is that intended? Do the `load_state_dict` below perform any operation?

Comment (ezyang) [reply torch/jit/__init__.py]: It's definitely intended to clone, so if this is not the case, we'll have to fix it and add a test.

File: torch/jit/__init__.py [hunk 2]
```diff
@@ -346,8 +340,12 @@ def __process_args(self, args):
 
     # NB: In principle, there could also be a 'raw' version of this compiler,
     # but since the logic is so complicated, testing code wouldn't benefit much
-    def __new_forward(self, *args):
-        if _JIT_DISABLE or not self.__enabled:
+    def __new_forward(self, *args, **kwargs):
+        enabled = kwargs.pop("enabled", True)
```
Comment (apaszke) [inline torch/jit/__init__.py]: This is really ugly. We shouldn't add any arguments to `forward`, especially `enabled=False`. Think that the complile annotation is very far from the actual use of the module, and `module(x, enabled=False)` looks very weird

File: torch/jit/__init__.py [hunk 3]
```diff
@@ -543,6 +517,19 @@ def _flatten(obj, params=tuple()):
     return obj_vars, obj_struct
 
 
+def _clone_inputs(args):
+    def clone_input(a):
+        if isinstance(a, Variable):
+            v = Variable(a.data.clone(), requires_grad=a.requires_grad, volatile=a.volatile)
+            if a.grad is not None:
+                v.grad = a.grad.clone()
```
Comment (apaszke) [inline torch/jit/__init__.py]: I'd rewrap the grad as well (+ can you even assign to `.grad`? I think you have to use `._grad` for this). If someone uses higher order grads, `.grad` is attached to a graph, and `clone()` will extend it

File: torch/jit/__init__.py [hunk 4]
```diff
@@ -580,5 +567,85 @@ def _time(trace_name, name, time=True):
...
 
 
+def verify(model, args, loss_fn=torch.sum, devices=None):
+    """
+    Verify that a JIT compiled model has the same behavior as its uncompiled
+    version along with its backwards pass.
+
+    This function has side-effects (e.g., it executes your model / saves and loads
+    parameters), so don't expect the model to come out exactly the same as what
+    you passed in.
```
Comment (apaszke) [inline torch/jit/__init__.py]: Executing a model and saving/loading parameters isn't an externally visible side effect. Let's just say that it **might have side effects if your model changes at every forward (e.g. you have BN in training mode)**. It might also be good switch the model to eval like we do with exports

File: torch/jit/__init__.py [hunk 5]
```diff
@@ -580,5 +567,85 @@ def _time(trace_name, name, time=True):
...
+
+    Arguments:
+        model (compiled torch.nn.Module): the compiled module/function to be
```
Comment (apaszke) [inline torch/jit/__init__.py]: we don't say anywhere that compiler replaces the thing with a module, right? Just say it's a compiled Module or function (in the type signature)

File: torch/jit/__init__.py [hunk 6]
```diff
@@ -580,5 +567,85 @@ def _time(trace_name, name, time=True):
...
+            `@torch.jit.compile`.
+        args (tuple or Variable): the positional arguments to pass to the
+            compiled function/module to be verified.  A non-tuple is assumed to
+            be a single positional argument to be passed to the model.
+        loss_fn (function, optional): the loss function to be applied to
+            the output of the model, before backwards is invoked.  By default,
+            we assume that a model returns a single result, and we :func:`torch.sum`
+            before calling backwards; if this is inappropriate, you can pass your
+            own loss function.  Note that if a model returns a tuple of results,
+            these are passed as separate arguments to `loss_fn`.
```
Comment (apaszke) [inline torch/jit/__init__.py]: I can't see why would summing the outputs would be inappropriate. It really doesn't matter and this argument is not needed. If you really don't want to sum them, just allocate a one-filled tensors as grad_output (this is equivalent)

File: torch/jit/__init__.py (hunk 1 above)
Comment (apaszke) [reply torch/jit/__init__.py]: It is a view. Do `copy.deepcopy` on it

File: torch/jit/__init__.py [hunk 7]
```diff
@@ -580,5 +567,85 @@ def _time(trace_name, name, time=True):
...
+    if not isinstance(model, _CompiledMixin):
+        raise TypeError("Cannot verify an uncompiled module.  Add @torch.jit.compile to compile it")
+
+    if not isinstance(args, tuple):
+        args = (args,)
+
+    verify_args = _clone_inputs(args)
+
+    verify_state = model.state_dict()
+    model.zero_grad()
```
Comment (apaszke) [inline torch/jit/__init__.py]: You don't need this. Use the functional API to compute all the gradients.

File: torch/jit/__init__.py [hunk 8]
```diff
@@ -580,5 +567,85 @@ def _time(trace_name, name, time=True):
...
+
+    # We use a special API to reset the trace and compile it from scratch.
+
+    with torch.random.fork_rng(devices, _caller="torch.jit.verify"):
+        out = model(*args, _force_trace=True)
```
Comment (apaszke) [inline torch/jit/__init__.py]: I'm not sure why do we need to force the trace here. Just "unwrap" the compiled module, use this to run the original forward, and then run the decorated thing twice (to make sure it has the trace)

File: torch/jit/__init__.py [hunk 9]
```diff
@@ -580,5 +567,85 @@ def _time(trace_name, name, time=True):
...
+            out = (out, )
+        out_vars, _ = _flatten(out)
+        saved_outs = [v.data.clone() for v in out_vars]
+        loss = loss_fn(*out)
+        loss.backward()
+        saved_grads = [v.grad.data.clone() for v in model.parameters()]
+        assert model.has_trace_for(*args)
+
+    model.load_state_dict(verify_state)
+    verify_out = model(*verify_args)
```
Comment (apaszke) [inline torch/jit/__init__.py]: `out` and `verify_out` give you absolutely no information about what they are. Which one is the original and which one is from the JIT?

File: torch/jit/__init__.py (hunk 2 above)
Comment (ezyang) [reply torch/jit/__init__.py]: The "enabled" is an accident. 

`_force_trace` is intended to be a private kwarg. If we really want to do it differently, I can alternately add a method to delete a trace associated with an argument.

File: torch/jit/__init__.py (hunk 3 above)
Comment (ezyang) [reply torch/jit/__init__.py]: The assignment is fine:

```
  {"_grad", (getter)THPVariable_get_grad, (setter)THPVariable_set_grad, NULL, NULL}, // only for legacy reasons 
  {"grad", (getter)THPVariable_get_grad, (setter)THPVariable_set_grad, NULL, NULL},
```

But I'll recursively call `clone_input` to handle `v.grad`.

File: torch/jit/__init__.py (hunk 4 above)
Comment (ezyang) [reply torch/jit/__init__.py]: So, the saving/loading parameters is done *solely* so that we can rerun the model with the compiled trace, but otherwise same initial conditions; we're not actually resetting the model to the very original parameter setting. We could do this (so that it has "no side effects") but there is more work to do; for example, you would have to save and restore the state of all the grads, and it just seems like complication for no good reason.

One thing to note is switching model to `eval` does NOT make sense, because a user may very well be interested in knowing that the model runs in training mode the same as compiled and as not.

File: torch/jit/__init__.py (hunk 6 above)
Comment (ezyang) [reply torch/jit/__init__.py]: This is more a question of user knobs. It would only be inappropriate if summing masked a problem with the network, which would be exposed with a different loss function. I don't know of any case where this could be true, but it seems better to let the user stick in their own loss function as needed.

File: torch/jit/__init__.py (hunk 8 above)
Comment (ezyang) [reply torch/jit/__init__.py]: The uncompiled class *never exists* with our current decorator strategy. Remember the discussion about `super(self, Foo)`: we only ever get to have `Foo` resolve to ONE class, and whichever class we send to, we're stuck with it.

File: torch/jit/__init__.py (hunk 6 above)
Comment (ezyang) [reply torch/jit/__init__.py]: Having read your comment about using `grad`, I now understand this comment: are you saying that there should be a `grad_outputs` input to `torch.jit.verify` instead of a `loss_fn`.

**Comment by ezyang** [comment]
Pushed with CI comment fixes

File: torch/jit/__init__.py (hunk 4 above)
Comment (apaszke) [reply torch/jit/__init__.py]: I'm confused. Is it saving as in `torch.save` or is it just clone? How is that a visible side effect? It's only executing the model that can mutate things

File: torch/jit/__init__.py (hunk 6 above)
Comment (apaszke) [reply torch/jit/__init__.py]: Then add an option to provide `grad_outputs`. That makes it easier to apply prepared inputs than to generate them via a loss function. We could keep both

File: torch/jit/__init__.py (hunk 4 above)
Comment (apaszke) [reply torch/jit/__init__.py]: BTW the default `loss_fn` will break as soon as someone has more than one output of the model. Might be worth fixing or at least mention that in the docs

File: torch/jit/__init__.py [hunk 10]
```diff
@@ -580,5 +570,77 @@ def _time(trace_name, name, time=True):
...
+        raise TypeError("Cannot verify an uncompiled module.  Add @torch.jit.compile to compile it")
+
+    if not isinstance(args, tuple):
+        args = (args,)
+
+    saved_args = _clone_inputs(args)
+    saved_state = copy.deepcopy(model.state_dict())
+
+    def run_fwd_bwd(args, force_trace=False):
+        in_vars, _ = _flatten(args, model.state_dict(keep_vars=True).values())
```
Comment (apaszke) [inline torch/jit/__init__.py]: `_flatten` drops all non-Variable objects, right (just want to be sure it doesn't break with buffers)?

File: torch/jit/__init__.py [hunk 11]
```diff
@@ -580,5 +570,77 @@ def _time(trace_name, name, time=True):
...
+        # We use a special API to reset the trace and compile it from scratch.
+        out = model(*args, _force_trace=force_trace)
+        if not isinstance(out, tuple):
+            out = (out, )
+        out_vars, _ = _flatten(out)
+        saved_outs = [v.data.clone() for v in out_vars]
+        loss = loss_fn(*out)
```
Comment (apaszke) [inline torch/jit/__init__.py]: You can't assume to do `*out` in general. If model outputs a tuple, you won't wrap it later, and you will unpack it in here.

File: torch/jit/__init__.py [hunk 12]
```diff
@@ -580,5 +570,77 @@ def _time(trace_name, name, time=True):
...
+        grads = torch.autograd.grad([loss], in_vars)
+        # TODO: I'm not sure if the clone here is necessary but it is safer
+        saved_grads = [v.data.clone() for v in grads]
+        return (saved_outs, saved_grads)
+
+    with torch.random.fork_rng(devices, _caller="torch.jit.verify"):
+        uncompiled_outs, uncompiled_grads = run_fwd_bwd(args, force_trace=True)
+        assert model.has_trace_for(*args)
+
+    model.load_state_dict(saved_state)
```
Comment (apaszke) [inline torch/jit/__init__.py]: Uh this is annoying. After `torch.jit.verify` param data will point to a different thing. Why can't you just use `copy` to create a clone of the whole module tree, including Parameters and use this to compare against the original module

**Review by apaszke — COMMENTED** [review]
Nearly good to merge. I just have a few last questions and I don't understand why do you do a few things this way.

File: torch/jit/__init__.py (hunk 12 above)
Comment (fmassa) [reply torch/jit/__init__.py]: I don't think this is true anymore. Since https://github.com/pytorch/pytorch/pull/451, `load_state_dict` [performs a `copy_` and not an assignment anymore](https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/module.py#L401), so I think this should be fine, right?

Comment (apaszke) [reply torch/jit/__init__.py]: Ah my bad. Thanks @fmassa

File: torch/jit/__init__.py (hunk 10 above)
Comment (ezyang) [reply torch/jit/__init__.py]: Yes, it appears buffers will get dropped. (Though, I suspect the entire buffer mechanism is broken right now...)

File: torch/jit/__init__.py (hunk 11 above)
Comment (ezyang) [reply torch/jit/__init__.py]: Actually, this was intentional: *if* the model returns a tuple, I want the loss function to handle it as a multivariadic function; if it returns just a single output, you just get one argument. It's more pleasant to write lambdas this way, because you can't destructure tuples in arguments.

File: torch/jit/__init__.py (hunk 12 above)
Comment (ezyang) [reply torch/jit/__init__.py]: To be honest, I wrote the code this way because I was not sure `copy` would always be supported on user-written Modules. If it could be broken in subtle ways, I would prefer not to use it, since it adds another way for things to fail (whereas `state_dict` and `load_state_dict` are blessed official APIs.) But if we do know deep copies always work, it is probably more robust to make a straight up copy at the beginning.
